File: src/main/resources/static/java/descending/QuickSortDescending.java
package java.descending;

import java.io.FileWriter;
import java.io.IOException;
import java.util.Random;
import java.util.Scanner;

public class QuickSortDescending {

    int partition(int arr[], int low, int high) {
        int pivot = arr[high];
        int i = low - 1; 

        for (int j = low; j < high; j++) {
            
            if (arr[j] >= pivot) {
                i++;
            
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    void quicksort(int arr[], int low, int high) {
        if (low < high) {
            
            int pivot = partition(arr, low, high);

            quicksort(arr, low, pivot - 1);
            quicksort(arr, pivot + 1, high);
        }
    }

    public static void main(String[] args) {
        QuickDescendingJava obj = new QuickDescendingJava();
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of the array: ");
        int k = sc.nextInt();

        int arr[] = new int[k];
        Random random = new Random();
        for (int i = 0; i < arr.length; i++) {
            arr[i] = random.nextInt(100) + 1;
        }

        long startTime = System.currentTimeMillis();

        String directory = "Sorting_Problems/";
        String originalFile = directory + "original_array.csv";
        String sortedFile = directory + "sorted_result.csv";

        try (FileWriter writer = new FileWriter(originalFile)) {
            for (int i = 0; i < arr.length; i++) {
                writer.append(String.valueOf(arr[i]));
                if (i < arr.length - 1) {
                    writer.append(",");
                }
            }
            writer.append("\n");
        } catch (IOException e) {
            e.printStackTrace();
        }

        obj.quicksort(arr, 0, arr.length - 1);

        try (FileWriter writer = new FileWriter(sortedFile)) {
            for (int i = 0; i < arr.length; i++) {
                writer.append(String.valueOf(arr[i]));
                if (i < arr.length - 1) {
                    writer.append(",");
                }
            }
            writer.append("\n");
        } catch (IOException e) {
            e.printStackTrace();
        }

        long endTime = System.currentTimeMillis();
        long totalTime = endTime - startTime;
        double seconds = totalTime / 1000.0; 

        System.out.println("Original and sorted arrays have been saved to CSV files.");
        System.out.println("Original array: " + originalFile);
        System.out.println("Sorted array: " + sortedFile);
        System.out.println("Time taken: " + seconds + " seconds.");

        sc.close();
    }
}

File: src/main/resources/static/cpp/descending/QuickSortDescending.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <ctime>
#include <cstdlib>
#include <string>
#include <filesystem>

// Function to perform Quick Sort
void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pivotIndex = low + (high - low) / 2;
        int pivot = arr[pivotIndex];
        int i = low, j = high;

        while (i <= j) {
            while (arr[i] > pivot) {
                i++;
            }
            while (arr[j] < pivot) {
                j--;
            }
            if (i <= j) {
                std::swap(arr[i], arr[j]);
                i++;
                j--;
            }
        }

        // Recursively sort the partitions
        if (low < j) {
            quickSort(arr, low, j);
        }
        if (i < high) {
            quickSort(arr, i, high);
        }
    }
}

// Function to write an array to a CSV file
void writeToFile(const std::string& filename, const std::vector<int>& arr) {
    std::ofstream file(filename);
    for (size_t i = 0; i < arr.size(); i++) {
        file << arr[i];
        if (i < arr.size() - 1) {
            file << ",";
        }
    }
    file << std::endl;
    file.close();
}

// Function to read an array from a CSV file
std::vector<int> readFromFile(const std::string& filename) {
    std::vector<int> arr;
    std::ifstream file(filename);
    std::string line;
    if (std::getline(file, line)) {
        size_t start = 0;
        size_t end = line.find(',');
        while (end != std::string::npos) {
            arr.push_back(std::stoi(line.substr(start, end - start)));
            start = end + 1;
            end = line.find(',', start);
        }
        arr.push_back(std::stoi(line.substr(start, end)));
    }
    return arr;
}

int main() {
    int n;
    // std::cout << "Enter the size of the array: ";
    std::cin >> n;

    // Get the current path and navigate to the static directory
    std::filesystem::path currentPath = std::filesystem::current_path();
    std::filesystem::path basePath = currentPath.parent_path().parent_path().parent_path() / "static";

    std::string originalFile = (basePath / "original_array.csv").string();
    std::string sortedFile = (basePath / "sorted_result.csv").string();

    std::vector<int> arr;

    // Always generate a new array based on user input
    arr.resize(n);
    std::srand(std::time(nullptr));
    for (int i = 0; i < n; i++) {
        arr[i] = std::rand() % 100 + 1;
    }
    writeToFile(originalFile, arr);

    clock_t startTime = clock();

    // Sort the array using Quick Sort
    quickSort(arr, 0, arr.size() - 1);

    // Write the sorted array to file
    writeToFile(sortedFile, arr);

    clock_t endTime = clock();
    double totalTime = static_cast<double>(endTime - startTime) / CLOCKS_PER_SEC;

    // // Output results
    // std::cout << "Original array file: " << originalFile << std::endl;
    // std::cout << "Sorted array file: " << sortedFile << std::endl;
    std::cout << "Time taken: " << totalTime << " seconds." << std::endl;

    return 0;
}
